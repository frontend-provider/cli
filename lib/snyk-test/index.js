module.exports = test;

var detect = require('../detect');
var runTest = require('./run-test');
var chalk = require('chalk');

function test(root, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  if (!options) {
    options = {};
  }

  var promise = executeTest(root, options);
  if (callback) {
    promise.then(function (res) {
      callback(null, res);
    }).catch(callback);
  }
  return promise;
}

function executeTest(root, options) {
  try {
    var packageManager = detect.detectPackageManager(root, options);
    options.packageManager = packageManager;
    return run(root, options)
      .then(function (res) {
        if (!res.packageManager) {
          res.packageManager = packageManager;
        }
        return res;
      });
  } catch (error) {
    return Promise.reject(chalk.red.bold(error));
  }
}

function run(root, options) {
  var packageManager = options.packageManager;
  if (['npm', 'yarn'].indexOf(packageManager) >= 0) {
    return require('./npm')(root, options);
  }
  if (!options.docker && [
    'rubygems',
    'maven',
    'gradle',
    'sbt',
    'pip',
    'golangdep',
    'govendor',
    'nuget',
    'composer',
  ].indexOf(packageManager) === -1) {
    throw new Error('Unsupported package manager: ' + packageManager);
  }
  return runTest(packageManager, root, options);
}
