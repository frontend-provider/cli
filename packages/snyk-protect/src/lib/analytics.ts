import { AnalyticsPayload, ProtectResult } from './types';
import { postJson } from './http';
import { getApiBaseUrl } from './snyk-api';
import * as os from 'os';
import { readFileSync, existsSync } from 'fs';
import * as path from 'path';

function getVersion(): string {
  return JSON.parse(
    readFileSync(path.resolve(__dirname, '../../package.json'), 'utf-8'),
  ).version;
}

function getAnalyticsData(result: ProtectResult): AnalyticsPayload {
  return {
    command: `@snyk/protect`,
    args: [],
    version: getVersion(),
    nodeVersion: process.version,
    metadata: {
      protectResult: result,
    },
  };
}

export async function sendAnalytics(result: ProtectResult) {
  if (!allowAnalytics()) {
    return;
  }
  try {
    const apiBaseUrl = getApiBaseUrl();
    const apiUrl = `${apiBaseUrl}/v1/analytics/cli`;
    const data = {
      data: getAnalyticsData(result),
    };
    await postJson(apiUrl, data);
  } catch (err) {
    // do nothing and log nothing
  }
}

function allowAnalytics(): boolean {
  try {
    const snykConfigFile = getSnykConfigFilePath();
    if (existsSync(snykConfigFile)) {
      const config = JSON.parse(readFileSync(snykConfigFile, 'utf-8'));
      if (
        config['disable-analytics'] === '1' ||
        config['disable-analytics'] === 1
      ) {
        return false;
      }
    }
    if (process.env.SNYK_DISABLE_ANALYTICS) {
      return false;
    }
  } catch (err) {
    // do nothing and log nothing
  }
  return true;
}

function getSnykConfigFilePath(): string {
  return (
    process.env.SNYK_CONFIG_FILE ||
    path.resolve(os.homedir(), '.config', 'configstore', 'snyk.json')
  );
}
