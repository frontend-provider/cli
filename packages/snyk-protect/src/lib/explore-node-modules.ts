import * as fs from 'fs';
import * as path from 'path';
import { FoundPhysicalPackage } from './types';

/**
 * Check if a physical module (given by folderPath) is one of the package names we are interested in for patching.
 * If it is, capture it as a FoundPhysicalPackage containg the name, version, and path
 * @param folderPath the physical path of the module under consideration.
 * @param packageNamesOfInterest a list of package names we are considering for patching.
 * @returns a FoundPhysicalPackage if the module at the given path has a package name which we are considering for patching.
 */
function checkPhysicalModule(
  folderPath: string,
  packageNamesOfInterest: Readonly<string[]>,
): FoundPhysicalPackage | undefined {
  const folderName = path.basename(folderPath);
  if (packageNamesOfInterest.includes(folderName)) {
    const packageJsonPath = path.resolve(folderPath, 'package.json');
    if (
      fs.existsSync(packageJsonPath) &&
      fs.lstatSync(packageJsonPath).isFile()
    ) {
      const { name, version } = JSON.parse(
        fs.readFileSync(packageJsonPath, 'utf8'),
      );
      if (packageNamesOfInterest.includes(name)) {
        const foundPackage: FoundPhysicalPackage = {
          packageName: name,
          packageVersion: version,
          path: folderPath,
        };
        return foundPackage;
      }
    }
  }
  return undefined;
}

/**
 * Explore the node_modules of a project, starting with the given path and recusively exploring deeper,
 * looking for modules which match the package names we are considering for patching.
 * @param pathToCheck the path to look for a matching module.
 * @param packageNamesOfInterest - a list of package names we are considering for patching.
 * @returns
 */
export function findPhysicalModules(
  pathToCheck: string,
  packageNamesOfInterest: Readonly<string[]>,
): FoundPhysicalPackage[] {
  if (fs.existsSync(pathToCheck) && fs.lstatSync(pathToCheck).isDirectory()) {
    const foundPhysicalPackages: FoundPhysicalPackage[] = [];
    const foundAtThisLocation = checkPhysicalModule(
      pathToCheck,
      packageNamesOfInterest,
    );
    if (foundAtThisLocation) {
      foundPhysicalPackages.push(foundAtThisLocation);
    }

    const folderNodeModules = path.resolve(pathToCheck, 'node_modules');
    if (
      fs.existsSync(folderNodeModules) &&
      fs.lstatSync(folderNodeModules).isDirectory()
    ) {
      fs.readdirSync(folderNodeModules).forEach((p) => {
        const foundPhysicalPackagesInSubModules = findPhysicalModules(
          path.resolve(folderNodeModules, p),
          packageNamesOfInterest,
        );
        foundPhysicalPackages.push(...foundPhysicalPackagesInSubModules);
      });
    }

    return foundPhysicalPackages;
  } else {
    return [];
  }
}
