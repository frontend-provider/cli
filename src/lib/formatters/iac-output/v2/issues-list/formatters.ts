import { FormattedResult } from '../../../../../cli/commands/test/iac/local-execution/types';
import { IacOutputMeta } from '../../../../types';
import { IacTestOutput } from './types';

export function formatScanResultsNewOutput(
  oldFormattedResults: FormattedResult[],
  outputMeta: IacOutputMeta,
): IacTestOutput {
  const newFormattedResults: IacTestOutput = {
    results: {},
    metadata: outputMeta,
  };

  oldFormattedResults.forEach((oldFormattedResult) => {
    oldFormattedResult.result.cloudConfigResults.forEach((issue) => {
      if (!newFormattedResults.results[issue.severity]) {
        newFormattedResults.results[issue.severity] = [];
      }

      newFormattedResults.results[issue.severity].push({
        issue,
        targetFile: oldFormattedResult.targetFile,
        projectType: oldFormattedResult.result.projectType,
      });
    });
  });

  return newFormattedResults;
}
