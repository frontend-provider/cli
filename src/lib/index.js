module.exports = snyk;

var cluster = require('cluster');
var snykConfig = require('./config');

// This module exports a function with properties attached.
// The properties represent other modules - so, it's some kind of "world object"
// that is used to indirectly import modules.
// The function modifies the object and returns it.
// This also introduces some circular imports.

// TODO(kyegupov): untangle this, resolve circular imports, convert to Typescript

function snyk(options) {
  if (!options) {
    options = {};
  }

  if (options.api) {
    snykConfig.api = options.api;
  }

  if (options.id) {
    snyk.id = options.id;
  }

  // FIXME add in pid + whether master + hostname + all these fields

  snyk.config.isMaster = cluster.isMaster;

  if (options.monitor && snyk.config.isMaster) {
    if (!snyk.api) {
      throw new Error('Snyk monitors require an authenticated account ' +
        'and API token');
    }
    // hook();
    require('./capture')();
  }

  return snyk;
}

snyk.id = snykConfig.id;

var apiToken = require('./api-token');

// make snyk.api *always* get the latest api token from the config store
Object.defineProperty(snyk, 'api', {
  enumerable: true,
  configurable: true,
  get: function () {
    return apiToken();
  },
  set: function (value) {
    snykConfig.api = value;
  },
});

snyk.modules = require('./modules');
snyk.test = require('./snyk-test');
snyk.bus = require('./bus');
snyk.policy = require('snyk-policy');
snyk.isRequired = true; // changed to false when loaded via cli
snyk.isolate = {
  okay: function () {
    return true;
  },
};

// this is the user config, and not the internal config
snyk.config = require('./user-config');
