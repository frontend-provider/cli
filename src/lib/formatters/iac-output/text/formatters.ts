import { FormattedResult } from '../../../../cli/commands/test/iac/local-execution/types';
import { iacRemediationTypes } from '../../../iac/constants';
import { Results, Vulnerability } from '../../../iac/test/v2/scan/results';
import { SEVERITY } from '../../../snyk-test/legacy';
import { IacOutputMeta } from '../../../types';
import {
  FormattedOutputResultsBySeverity,
  IacTestCounts,
  IacTestData,
  Issue,
} from './types';
import { countSuppressedIssues } from './utils';

interface FormatTestDataParams {
  oldFormattedResults: FormattedResult[];
  iacOutputMeta: IacOutputMeta | undefined;
  ignoresCount: number;
}

export function formatTestData({
  oldFormattedResults,
  iacOutputMeta: iacTestMeta,
  ignoresCount,
}: FormatTestDataParams): IacTestData {
  const resultsBySeverity = formatScanResultsNewOutput(oldFormattedResults);

  return {
    resultsBySeverity,
    metadata: iacTestMeta,
    counts: formatTestCounts(resultsBySeverity, {
      oldFormattedResults,
      ignoresCount,
    }),
  };
}

function formatTestCounts(
  resultsBySeverity: FormattedOutputResultsBySeverity,
  {
    oldFormattedResults,
    ignoresCount,
  }: Pick<FormatTestDataParams, 'ignoresCount' | 'oldFormattedResults'>,
): IacTestCounts {
  const filesWithIssues = oldFormattedResults.filter(
    (result) => result.result.cloudConfigResults.length,
  ).length;

  const filesWithoutIssues = oldFormattedResults.length - filesWithIssues;

  let totalIssues = 0;

  const issuesCountBySeverity = {} as { [key in SEVERITY]: number };
  Object.values(SEVERITY).forEach((severity) => {
    issuesCountBySeverity[severity] = resultsBySeverity[severity]?.length || 0;
    totalIssues += issuesCountBySeverity[severity];
  });

  return {
    ignores: ignoresCount,
    filesWithIssues,
    filesWithoutIssues,
    issuesBySeverity: issuesCountBySeverity,
    issues: totalIssues,
  };
}

function formatScanResultsNewOutput(
  oldFormattedResults: FormattedResult[],
): FormattedOutputResultsBySeverity {
  const newFormattedResults: FormattedOutputResultsBySeverity = {};

  oldFormattedResults.forEach((oldFormattedResult) => {
    oldFormattedResult.result.cloudConfigResults.forEach((issue) => {
      if (!newFormattedResults[issue.severity]) {
        newFormattedResults[issue.severity] = [];
      }

      newFormattedResults[issue.severity].push({
        issue,
        targetFile: oldFormattedResult.targetFile,
        projectType: oldFormattedResult.result.projectType,
      });
    });
  });

  return newFormattedResults;
}

export function formatSnykIacTestTestData(
  snykIacTestScanResult: Results | undefined,
  projectName: string,
  orgName: string,
): IacTestData {
  const resultsBySeverity = formatSnykIacTestScanResultNewOutput(
    snykIacTestScanResult,
  );

  let totalIssues = 0;

  const issuesCountBySeverity = {} as { [key in SEVERITY]: number };
  Object.values(SEVERITY).forEach((severity) => {
    issuesCountBySeverity[severity] = resultsBySeverity[severity]?.length || 0;
    totalIssues += issuesCountBySeverity[severity];
  });

  const allFilesCount = countFiles(snykIacTestScanResult);
  const filesWithIssuesCount = countFilesWithIssues(snykIacTestScanResult);
  const filesWithoutIssuesCount = allFilesCount - filesWithIssuesCount;
  const ignores = snykIacTestScanResult
    ? snykIacTestScanResult.metadata.ignoredCount
    : 0;

  let contextSuppressedIssueCount: number | undefined;
  const suppressedResults =
    snykIacTestScanResult?.scanAnalytics?.suppressedResults;
  if (suppressedResults) {
    contextSuppressedIssueCount = countSuppressedIssues(suppressedResults);
  }

  return {
    resultsBySeverity,
    metadata: { projectName, orgName },
    counts: {
      ignores,
      filesWithIssues: filesWithIssuesCount,
      filesWithoutIssues: filesWithoutIssuesCount,
      issues: totalIssues,
      issuesBySeverity: issuesCountBySeverity,
      contextSuppressedIssues: contextSuppressedIssueCount,
    },
  };
}

function countFilesWithIssues(results?: Results): number {
  if (results && results.vulnerabilities) {
    const files = new Set<string>();

    for (const vulnerability of results.vulnerabilities) {
      if (vulnerability.resource.file) {
        files.add(vulnerability.resource.file);
      }
    }

    return files.size;
  }

  return 0;
}

function countFiles(results?: Results): number {
  if (results && results?.resources) {
    const files = new Set<string>();

    for (const resource of results.resources) {
      if (resource.file) {
        files.add(resource.file);
      }
    }

    return files.size;
  }

  return 0;
}

function formatSnykIacTestScanResultNewOutput(
  snykIacTestScanResult: Results | undefined,
): FormattedOutputResultsBySeverity {
  const resultsBySeverity = {} as FormattedOutputResultsBySeverity;

  if (snykIacTestScanResult?.vulnerabilities) {
    snykIacTestScanResult.vulnerabilities.forEach((vulnerability) => {
      if (!resultsBySeverity[vulnerability.severity]) {
        resultsBySeverity[vulnerability.severity] = [];
      }

      resultsBySeverity[vulnerability.severity]!.push({
        issue: formatSnykIacTestScanVulnerability(vulnerability),
        targetFile: vulnerability.resource.file,
        projectType: vulnerability.resource.kind,
      });
    });
  }

  return resultsBySeverity;
}

function formatSnykIacTestScanVulnerability(
  vulnerability: Vulnerability,
): Issue {
  const resolve = extractResolve(vulnerability);

  return {
    id: vulnerability.rule.id,
    severity: vulnerability.severity,
    title: vulnerability.rule.title,
    lineNumber: vulnerability.resource.line ?? -1,
    cloudConfigPath: formatCloudConfigPath(vulnerability),
    issue: vulnerability.rule.title,
    impact: vulnerability.rule.description,
    resolve,
    documentation: vulnerability.rule.documentation,
    remediation: {
      [iacRemediationTypes[vulnerability.resource.kind]]: resolve,
    },
  };
}
function extractResolve(vulnerability: Vulnerability): string {
  const newLineIdx = vulnerability.remediation.search(/\r?\n|\r/g);
  return newLineIdx < 0
    ? vulnerability.remediation
    : vulnerability.remediation.substring(0, newLineIdx);
}

function formatCloudConfigPath(vulnerability: Vulnerability): string[] {
  const cloudConfigPath = vulnerability.resource.id.split('.');

  if (vulnerability.resource.path) {
    cloudConfigPath.push(...vulnerability.resource.path);
  }

  return cloudConfigPath;
}
