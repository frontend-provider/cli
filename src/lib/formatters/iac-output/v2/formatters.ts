import { FormattedResult } from '../../../../cli/commands/test/iac/local-execution/types';
import { Results, Vulnerability } from '../../../iac/test/v2/scan/results';
import { AnnotatedIacIssue } from '../../../snyk-test/iac-test-result';
import { SEVERITY } from '../../../snyk-test/legacy';
import { IacOutputMeta } from '../../../types';
import {
  FormattedOutputResultsBySeverity,
  IacTestCounts,
  IacTestData,
} from './types';

interface FormatTestDataParams {
  oldFormattedResults: FormattedResult[];
  iacOutputMeta: IacOutputMeta | undefined;
  ignoresCount: number;
}

export function formatTestData({
  oldFormattedResults,
  iacOutputMeta: iacTestMeta,
  ignoresCount,
}: FormatTestDataParams): IacTestData {
  const resultsBySeverity = formatScanResultsNewOutput(oldFormattedResults);

  return {
    resultsBySeverity,
    metadata: iacTestMeta,
    counts: formatTestCounts(resultsBySeverity, {
      oldFormattedResults,
      ignoresCount,
    }),
  };
}

function formatTestCounts(
  resultsBySeverity: FormattedOutputResultsBySeverity,
  {
    oldFormattedResults,
    ignoresCount,
  }: Pick<FormatTestDataParams, 'ignoresCount' | 'oldFormattedResults'>,
): IacTestCounts {
  const filesWithIssues = oldFormattedResults.filter(
    (result) => result.result.cloudConfigResults.length,
  ).length;

  const filesWithoutIssues = oldFormattedResults.length - filesWithIssues;

  let totalIssues = 0;

  const issuesCountBySeverity = {} as { [key in SEVERITY]: number };
  Object.values(SEVERITY).forEach((severity) => {
    issuesCountBySeverity[severity] = resultsBySeverity[severity]?.length || 0;
    totalIssues += issuesCountBySeverity[severity];
  });

  return {
    ignores: ignoresCount,
    filesWithIssues,
    filesWithoutIssues,
    issuesBySeverity: issuesCountBySeverity,
    issues: totalIssues,
  };
}

function formatScanResultsNewOutput(
  oldFormattedResults: FormattedResult[],
): FormattedOutputResultsBySeverity {
  const newFormattedResults: FormattedOutputResultsBySeverity = {};

  oldFormattedResults.forEach((oldFormattedResult) => {
    oldFormattedResult.result.cloudConfigResults.forEach((issue) => {
      if (!newFormattedResults[issue.severity]) {
        newFormattedResults[issue.severity] = [];
      }

      newFormattedResults[issue.severity].push({
        issue,
        targetFile: oldFormattedResult.targetFile,
        projectType: oldFormattedResult.result.projectType,
      });
    });
  });

  return newFormattedResults;
}

export function formatSnykIacTestTestData(
  snykIacTestScanResult: Results | undefined,
  projectName: string,
  orgName: string,
): IacTestData {
  const resultsBySeverity = formatSnykIacTestScanResultNewOutput(
    snykIacTestScanResult,
  );

  let totalIssues = 0;

  const issuesCountBySeverity = {} as { [key in SEVERITY]: number };
  Object.values(SEVERITY).forEach((severity) => {
    issuesCountBySeverity[severity] = resultsBySeverity[severity]?.length || 0;
    totalIssues += issuesCountBySeverity[severity];
  });

  const allFilesCount = countFiles(snykIacTestScanResult);
  const filesWithIssuesCount = countFilesWithIssues(snykIacTestScanResult);
  const filesWithoutIssuesCount = allFilesCount - filesWithIssuesCount;

  return {
    resultsBySeverity,
    metadata: { projectName, orgName },
    counts: {
      ignores: 0,
      filesWithIssues: filesWithIssuesCount,
      filesWithoutIssues: filesWithoutIssuesCount,
      issues: totalIssues,
      issuesBySeverity: issuesCountBySeverity,
    },
  };
}

function countFilesWithIssues(results?: Results): number {
  if (results && results.vulnerabilities) {
    const files = new Set<string>();

    for (const vulnerability of results.vulnerabilities) {
      if (vulnerability.resource.file) {
        files.add(vulnerability.resource.file);
      }
    }

    return files.size;
  }

  return 0;
}

function countFiles(results?: Results): number {
  if (results && results?.resources) {
    const files = new Set<string>();

    for (const resource of results.resources) {
      if (resource.file) {
        files.add(resource.file);
      }
    }

    return files.size;
  }

  return 0;
}

function formatSnykIacTestScanResultNewOutput(
  snykIacTestScanResult: Results | undefined,
): FormattedOutputResultsBySeverity {
  const resultsBySeverity = {} as FormattedOutputResultsBySeverity;

  if (snykIacTestScanResult?.vulnerabilities) {
    snykIacTestScanResult.vulnerabilities.forEach((vulnerability) => {
      if (!resultsBySeverity[vulnerability.severity]) {
        resultsBySeverity[vulnerability.severity] = [];
      }

      resultsBySeverity[vulnerability.severity]!.push({
        issue: formatSnykIacTestScanVulnerability(vulnerability),
        targetFile: vulnerability.resource.file,
        projectType: vulnerability.resource.type,
      });
    });
  }

  return resultsBySeverity;
}

function formatSnykIacTestScanVulnerability(
  vulnerability: Vulnerability,
): AnnotatedIacIssue {
  return {
    id: vulnerability.rule.id,
    severity: vulnerability.severity,
    title: vulnerability.rule.title,
    isIgnored: vulnerability.ignored,
    lineNumber: vulnerability.resource.line,
    cloudConfigPath: vulnerability.resource.id
      .split('.')
      .concat(vulnerability.resource.path as string[]),
    subType: '',
    iacDescription: {
      issue: '',
      impact: '',
      resolve: '',
    },
  };
}
