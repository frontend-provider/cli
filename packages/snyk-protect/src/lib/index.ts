import * as fs from 'fs';
import * as path from 'path';
import { extractPatchMetadata } from './snyk-file';
import { applyPatchToFile } from './patch';
import { findPhysicalModules } from './explore-node-modules';
import {
  VulnIdAndPackageName,
  VulnPatches,
  PatchedModule,
  ProtectResultType,
} from './types';
import { getAllPatches } from './fetch-patches';
import { sendAnalytics } from './analytics';

async function protect(projectFolderPath: string) {
  const snykFilePath = path.resolve(projectFolderPath, '.snyk');

  if (!fs.existsSync(snykFilePath)) {
    console.log('No .snyk file found');
    sendAnalytics({
      type: ProtectResultType.NO_SNYK_FILE,
    });
    return;
  }

  const snykFileContents = fs.readFileSync(snykFilePath, 'utf8');
  const snykFilePatchMetadata = extractPatchMetadata(snykFileContents);
  const vulnIdAndPackageNames: VulnIdAndPackageName[] = snykFilePatchMetadata;
  const targetPackageNames = [
    ...new Set(snykFilePatchMetadata.map((vpn) => vpn.packageName)), // get a list of unique package names by converting to Set and then back to array
  ];

  // find instances of the target packages by spelunking through the node_modules looking for modules with a target packageName
  const foundPhysicalPackages = findPhysicalModules(
    projectFolderPath,
    targetPackageNames,
  );

  // Map of package name to versions (for the target package names).
  // For each package name, we might have found multiple versions and we'll need to fetch patches for each version.
  // We will use this to lookup the versions of packages to get patches for.
  const packageNameToVersionsMap = new Map<string, string[]>();
  foundPhysicalPackages.forEach((p) => {
    if (packageNameToVersionsMap.has(p.packageName)) {
      const versions = packageNameToVersionsMap.get(p.packageName);
      if (!versions?.includes(p.packageVersion)) {
        versions?.push(p.packageVersion);
      }
    } else {
      packageNameToVersionsMap.set(p.packageName, [p.packageVersion]);
    }
  });

  const packageAtVersionsToPatches: Map<
    string,
    VulnPatches[]
  > = await getAllPatches(vulnIdAndPackageNames, packageNameToVersionsMap);

  if (packageAtVersionsToPatches.size === 0) {
    console.log('Nothing to patch');
    sendAnalytics({
      type: ProtectResultType.NOTHING_TO_PATCH,
    });
    return;
  }

  const patchedModules: PatchedModule[] = [];
  foundPhysicalPackages.forEach((fpp) => {
    const packageNameAtVersion = `${fpp.packageName}@${fpp.packageVersion}`;
    const vuldIdAndPatches = packageAtVersionsToPatches.get(
      packageNameAtVersion,
    );
    vuldIdAndPatches?.forEach((vp) => {
      vp.patches.forEach((patchDiffs) => {
        patchDiffs.patchDiffs.forEach((diff) => {
          applyPatchToFile(diff, fpp.path);
        });
      });
      patchedModules.push({
        vulnId: vp.vulnId,
        packageName: fpp.packageName,
        packageVersion: fpp.packageVersion,
      });
    });
  });

  console.log('Successfully applied Snyk patches');

  sendAnalytics({
    type: ProtectResultType.APPLIED_PATCHES,
    patchedModules,
  });
}

export default protect;
