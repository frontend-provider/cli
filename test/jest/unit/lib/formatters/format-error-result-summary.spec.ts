import stripAnsi from 'strip-ansi';

import { summariseErrorResults } from '../../../../../src/lib/formatters';

describe('summariseErrorResults', () => {
  it('Single test result', () => {
    expect(stripAnsi(summariseErrorResults(1))).toMatchSnapshot();
  });

  it('Multiple test results', () => {
    expect(stripAnsi(summariseErrorResults(4))).toMatchSnapshot();
  });
});
