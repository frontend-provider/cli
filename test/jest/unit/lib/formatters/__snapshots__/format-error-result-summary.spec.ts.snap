// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`summariseErrorResults Multiple test results 1`] = `
" Failed to test 4 projects.
Run with \`-d\` for debug output and contact support@snyk.io"
`;

exports[`summariseErrorResults Single test result 1`] = `
" Failed to test 1 project.
Run with \`-d\` for debug output and contact support@snyk.io"
`;
